File: examples/governance/compliance_domain/lib/compliance_domain/training_record/commands/assign_training.rb
module ComplianceDomain
  class TrainingRecord
    module Commands
      class AssignTraining
        include Hecks::Command
        emits "AssignedTraining"

        attr_reader :stakeholder_id, :policy_id

        def initialize(stakeholder_id: nil, policy_id: nil)
          @stakeholder_id = stakeholder_id
          @policy_id = policy_id
        end

        def call
          TrainingRecord.new(
            stakeholder_id: stakeholder_id,
            policy_id: policy_id,
            status: "assigned"
          )
        end
      end
    end
  end
end

File: examples/governance_static_go/domain/assign_training.go
package domain

import (
	"time"
)

type AssignTraining struct {
	StakeholderId string `json:"stakeholder_id"`
	PolicyId string `json:"policy_id"`
}

func (c AssignTraining) CommandName() string { return "AssignTraining" }

func (c AssignTraining) Execute(repo TrainingRecordRepository) (*TrainingRecord, *AssignedTraining, error) {
	agg := NewTrainingRecord(c.StakeholderId, c.PolicyId, time.Time{}, time.Time{}, "", "")
	agg.Status = "assigned"
	if err := agg.Validate(); err != nil {
		return nil, nil, err
	}
	if err := repo.Save(agg); err != nil {
		return nil, nil, err
	}
	event := AssignedTraining{
		AggregateID: agg.ID,
		StakeholderId: c.StakeholderId,
		PolicyId: c.PolicyId,
		OccurredAt: time.Now(),
	}
	return agg, &event, nil
}
